diff --git a/src/exports/vite/index.js b/src/exports/vite/index.js
index 4a441b23d4d54ee10dd093733a4b2b60ab0742db..ed7978193f8cdb03a71b4f04afe89e8f7f25df72 100644
--- a/src/exports/vite/index.js
+++ b/src/exports/vite/index.js
@@ -1023,7 +1023,7 @@ async function kit({ svelte_config }) {
 					app_path: `${kit.paths.base.slice(1)}${kit.paths.base ? '/' : ''}${kit.appDir}`,
 					manifest_data,
 					out_dir: out,
-					service_worker: service_worker_entry_file ? 'service-worker.js' : null, // TODO make file configurable?
+					service_worker: 'service-worker.js',
 					client: null,
 					server_manifest
 				};
diff --git a/src/runtime/server/page/render.js b/src/runtime/server/page/render.js
index c3905d035b84b3664ae86aca6c1094c69df0bcd6..5f35ee12276f129f0648c9e24794c44108c990fe 100644
--- a/src/runtime/server/page/render.js
+++ b/src/runtime/server/page/render.js
@@ -536,7 +536,7 @@ export async function render_response({
 			blocks.push(boot);
 		}
 
-		if (options.service_worker) {
+		if (true) {
 			let opts = DEV ? ", { type: 'module' }" : '';
 			if (options.service_worker_options != null) {
 				const service_worker_options = { ...options.service_worker_options };
@@ -550,7 +550,11 @@ export async function render_response({
 			// older browsers (https://github.com/sveltejs/kit/pull/5417)
 			blocks.push(`if ('serviceWorker' in navigator) {
 						addEventListener('load', function () {
-							navigator.serviceWorker.register('${prefixed('service-worker.js')}'${opts});
+							navigator.serviceWorker.register('${prefixed('service-worker.js')}'${opts}).then(function (registration) {
+								if (registration.active) {
+									registration.active.postMessage({ __SVELTEKIT_VERCEL_SKEW_PROTECTION__: "${process.env.VERCEL_DEPLOYMENT_ID}" });
+								}
+							});
 						});
 					}`);
 		}
