diff --git a/src/exports/vite/index.js b/src/exports/vite/index.js
index 4a441b23d4d54ee10dd093733a4b2b60ab0742db..ed7978193f8cdb03a71b4f04afe89e8f7f25df72 100644
--- a/src/exports/vite/index.js
+++ b/src/exports/vite/index.js
@@ -1023,7 +1023,7 @@ async function kit({ svelte_config }) {
 					app_path: `${kit.paths.base.slice(1)}${kit.paths.base ? '/' : ''}${kit.appDir}`,
 					manifest_data,
 					out_dir: out,
-					service_worker: service_worker_entry_file ? 'service-worker.js' : null, // TODO make file configurable?
+					service_worker: 'service-worker.js',
 					client: null,
 					server_manifest
 				};
